idaes/apps/grid_integration/examples/thermal_generator_storage.py

```diff
@@ -0,0 +1,1069 @@
+                        round_trip_eff, \
+                        storage_size_hour,\
+                        storage_pmax_ratio,\
+                        rts_gmlc_data_dir,\
+                        **kwargs):
+
+    n_none = _check_number_of_None([storage_size_ratio,storage_size_hour,storage_pmax_ratio])
+
+    assert n_none == 1, \
+    'Please provide EXACTLY 2 of the 3 elements: storage capacity (storage_size_ratio), storage size in hour, and storage Pmax!'
+
+    # read data
+    gen_params = pd.read_csv(rts_gmlc_data_dir + 'gen.csv')
+    gen_params = gen_params[gen_params['GEN UID'].isin(generator_names)]
+
+    model_data = {}
+
+    # generator names
+    model_data['Generator'] = generator_names
+
+    # Pmin [MW]
+    model_data['Pmin'] = df_col_to_dict(df = gen_params,\
+                                        key_col_name = 'GEN UID',\
+                                        value_col_name = 'PMin MW')
+
+    # Pmax [MW]
+    model_data['Pmax'] = df_col_to_dict(df = gen_params,\
+                                        key_col_name = 'GEN UID',\
+                                        value_col_name = 'PMax MW')
+
+    # minimum up time [MW/hr]
+    model_data['UT'] = dict(zip(generator_names,gen_params['Min Up Time Hr'].astype(int)))
+
+    # minimum down time [hr]
+    model_data['DT'] = dict(zip(generator_names,gen_params['Min Down Time Hr'].astype(int)))
+
+    ## ramp rates [MW/hr]
+    ramp_rates = gen_params['Ramp Rate MW/Min'].values * 60
+
+    # ramp up rate [MW/hr]
+    model_data['RU'] = dict(zip(generator_names,ramp_rates))
+
+    # ramp down rate [MW/hr]
+    model_data['RD'] = dict(zip(generator_names,ramp_rates))
+
+    # ramp start up [MW/hr]
+    model_data['SU'] = {gen: min(model_data['Pmin'][gen],model_data['RU'][gen]) for gen in generator_names}
+
+    # ramp shut down [MW/hr] (will use pmin for now)
+    model_data['SD'] = {gen: min(model_data['Pmin'][gen],model_data['RD'][gen]) for gen in generator_names}
+
+    # start up cost [$/SU] (will use the warm start up cost for now)
+    start_up_cost = gen_params['Start Heat Warm MBTU'] * gen_params['Fuel Price $/MMBTU']
+    model_data['SU Cost'] = dict(zip(generator_names,start_up_cost))
+
+    ## production cost
+
+    # power segments and marginal costs
+    model_data['Power Segments'] = {}
+    model_data['Marginal Costs'] = {}
+    for gen in generator_names:
+        df = get_data_given(df = gen_params, generator = gen)
+        for l in range(1,4):
+            # power segements
+            model_data['Power Segments'][(gen,l)] = float(df['Output_pct_{}'.format(l)] * df['PMax MW'])
+
+            # segment marginal cost
+            model_data['Marginal Costs'][(gen,l)] = float(df['HR_incr_{}'.format(l)]/1000 * df['Fuel Price $/MMBTU'])
+
+    model_data['Min Load Cost'] = dict(zip(generator_names,gen_params['HR_avg_0']/1000 * gen_params['Fuel Price $/MMBTU'] * gen_params['PMin MW']))
+
+    if storage_size_ratio is None:
```

Comment:
Add a sentence comment after each part of this else/if block to explain what is happening.